File: services/ai/src/components/dqn/dqn.ts
import * as Tf from "@tensorflow/tfjs-node";

export	class	Dqn {
	public static ft_createDeepQNetwork(h: number, w: number, numActions: number): Tf.Sequential {
		if (h < 0 || w < 0 || numActions < 1) {
			throw new Error(`h: ${h}, w: ${w}, or numActions: ${numActions} is wrong.`);
		}
		const model = Tf.sequential();
		model.add(Tf.layers.conv2d({
			filters: 128,
			kernelSize: 3,
			strides: 1,
			activation: 'relu',
			inputShape: [h, w, 3]
		}));
		model.add(Tf.layers.batchNormalization());
		model.add(Tf.layers.conv2d({
			filters: 256,
			kernelSize: 3,
			strides: 1,
			activation: 'relu'
		}));
		model.add(Tf.layers.batchNormalization());
		model.add(Tf.layers.conv2d({
			filters: 256,
			kernelSize: 3,
			strides: 1,
			activation: 'relu'
		}));
		model.add(Tf.layers.flatten());
		model.add(Tf.layers.dense({units: 100, activation: 'relu'}));
		model.add(Tf.layers.dropout({rate: 0.25}));
		model.add(Tf.layers.dense({units: numActions}));
		return model;
	}

	public	static ft_copyWeights(destNetwork: Tf.Sequential, srcNetwork: Tf.Sequential): void {
		let originalDestNetworkTrainable;
		if (destNetwork.trainable !== srcNetwork.trainable) {
			originalDestNetworkTrainable = destNetwork.trainable;
			destNetwork.trainable = srcNetwork.trainable;
		}
		destNetwork.setWeights(srcNetwork.getWeights());
		if (originalDestNetworkTrainable != null) {
			destNetwork.trainable = originalDestNetworkTrainable;
		}
	}
}

File: services/ai_python/src_py/components/dqn/dqn.py
import tensorflow as tf

class Dqn:
    @staticmethod
    def ft_createDeepQNetwork(h: int, w: int, num_actions: int) -> tf.keras.Sequential:
        if h < 0 or w < 0 or num_actions < 1:
            raise ValueError(f"h:{h}, w:{w}, num_actions:{num_actions} invalid")
        model = tf.keras.Sequential([
            tf.keras.layers.Conv2D(128, 3, strides=1, activation='relu', input_shape=(h, w, 3)),
            tf.keras.layers.BatchNormalization(),
            tf.keras.layers.Conv2D(256, 3, strides=1, activation='relu'),
            tf.keras.layers.BatchNormalization(),
            tf.keras.layers.Conv2D(256, 3, strides=1, activation='relu'),
            tf.keras.layers.Flatten(),
            tf.keras.layers.Dense(100, activation='relu'),
            tf.keras.layers.Dropout(0.25),
            tf.keras.layers.Dense(num_actions)
        ])
        return model

    @staticmethod
    def ft_copyWeights(destNetwork: tf.keras.Model, srcNetwork: tf.keras.Model) -> None:
        original = destNetwork.trainable
        if destNetwork.trainable != srcNetwork.trainable:
            destNetwork.trainable = srcNetwork.trainable
        destNetwork.set_weights(srcNetwork.get_weights())
        destNetwork.trainable = original
